common/rpc/grpc.go

```diff
@@ -181,20 +185,51 @@ func NewServiceErrorInterceptor(
 		}
 		return resp, serviceerror.ToStatus(err).Err()
 	}
+}
+
+func NewFrontendErrorInterceptor(
+	logger log.Logger, hideErrorsAtFrontend bool,
+) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		_ *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
 
+		resp, err := handler(ctx, req)
+		if err != nil && hideErrorsAtFrontend {
+			err = HideUnknownOrInternalErrors(logger, err)
+		}
+		return resp, serviceerror.ToStatus(err).Err()
+	}
 }
 
-func FrontendErrorInterceptor(
-	ctx context.Context,
-	req interface{},
-	_ *grpc.UnaryServerInfo,
-	handler grpc.UnaryHandler,
-) (interface{}, error) {
-	resp, err := handler(ctx, req)
+var errorFrontendUnknown = errors.New("something went wrong, please retry")
 
-	// mask some internal errors at frontend
+func HideUnknownOrInternalErrors(
+	logger log.Logger,
+	err error,
+) error {
+	// convert some internal errors into specific errors
 	if _, ok := err.(*serviceerrors.ShardOwnershipLost); ok {
-		err = serviceerror.NewUnavailable("shard unavailable, please backoff and retry")
+		return serviceerror.NewUnavailable("shard unavailable, please backoff and retry")
 	}
-	return resp, serviceerror.ToStatus(err).Err()
+
+	s := serviceerror.ToStatus(err)
+	if s.Code() != codes.Unknown && s.Code() != codes.Internal {
+		return err
+	}
+
+	// convert internal and unknown errors into neutral error with hash code of the original error
+	errorHash := errorMessageHash(err)
+	// log original error with hash
+	message := fmt.Sprintf("%s (%s)", err.Error(), errorHash)
+	logger.Error(message)
+
+	return fmt.Errorf("%w (%s)", errorFrontendUnknown, errorHash)
```

Comment:
I think we should return a defined error type from serviceerror. I think Unavailable is appropriate here.

Reply:
this way we are changing real error type. I will return the original error code (Unknown or Internal).